get_prism_data.R

```diff
@@ -4,39 +4,232 @@ library(prism)
+    locations <- unique(select(bbs_data, site_id, long, lat))
+    coordinates(locations) <- c("long", "lat")
+    
+    #Check to see if all the raw data in the years specified are downloaded,
+    #download everything again if not. (Getting only what is needed, say if a 
+    #previous download failed, might be overly complicated)
+    if(!check_If_Prism_Files_Present(ls_prism_data(), yearsToUse)){
+      downloadPrism()
+    }
+    
+    #Load the prism data and extract using the bbs locations.
+    prism_stacked <- prism_stack(ls_prism_data())
+    extracted <- raster::extract(prism_stacked, locations)
+    prism_bbs_data <- data.frame(site_id = locations$site_id, coordinates(locations), extracted)
+    prism_bbs_data <- prism_bbs_data %>%
+      gather(date, value, 4:ncol(prism_bbs_data)) %>%
+      tidyr::extract(date, c("clim_var", "year", "month"),
+                     "PRISM_([:alpha:]*)_stable_[:alnum:]*_([:digit:]{4})([:digit:]{2})_")
+    
+    #Format the data a little and load into the sqlite database.
+    prism_bbs_data$year <- as.numeric(prism_bbs_data$year)
+    prism_bbs_data$month <- as.numeric(prism_bbs_data$month)
+    mydata <- copy_to(database, prism_bbs_data, temporary = FALSE,
+                      indexes = list(c("site_id", "year", "month")))
+    
+    #Now return the data as asked for
+    return(prism_bbs_data)
+    
+  } else {
+    
+    stop('Im stopping because the sqlite query for prism_bbs data returned neither a data frame nor an error in get_prism_data()')
+  }
+  
+}
+
+
+#################################################################
+#From raw prism monthly values calculate all the bioclim variables.
+#You should not call this directly to load bioclim vars. Instead call
+#get_bioclim_data(), which will 1st try to load the data from the sdqlite
+#db before processing it all from scratch. 
+###################################################################
+#First a quick helper function to calculate some of the bioclim variables,
+#like "precip in coldest month"
+maxMinCombo=function(vec1,vec2,max=TRUE){
+  #Return the value in vec1 in the position where
+  #vec2 is either highest or lowest. But 1st check for na 
+  #values. 
+  if(any(is.na(vec1)) | any(is.na(vec2))){
+    return(NA)
+  } else  if(max){
+    return(vec1[which.max(vec2)])
+  } else {
+    return(vec1[which.min(vec2)])
   }
 }
 
-datadirs = dir(datapath)
-for (datadir in datadirs) {
-  bil_file = paste(datadir, '.bil', sep = "")
-  bil_file_path = file.path(datapath, datadir, bil_file)
-  sql_file = paste(datadir, ".sql", sep = "")
-  sql_file_path = file.path(datapath, datadir, sql_file)
-  system(paste("raster2pgsql -s 4326", bil_file_path, datadir, ">", sql_file_path))
-  system(paste("psql -d bbsforecasting -f", sql_file_path))
+process_bioclim_data=function(){
+  #Get the prism data. It's pulled from the sqlite DB or downloaded as needed.
```

Comment:
Comments like this are generally considered unnecessary because they just restate the information that you get by reading the code. If folks need to know where the PRISM data is coming from they can read `get_prism_data()`.